### PR TITLE
Add telegram notification support

### PR DESCRIPTION
**Pull request by sohm22**
- new flag to choose notifier between email and telegram  ex :- `--notifier-type telegram` 
- add other flags to add telegram details `--telegram-token  --telegram-username`
- add telegram bot initialization and send notification functionality
- update notification workflow
- create document for telegram bot integration with covaccine-notifier
- send telegram notification as file if message length exceeds 4096 characters
- this PR addresses the issue #24

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -55,6 +59,9 @@ func init() {
 	rootCmd.PersistentFlags().IntVarP(&interval, "interval", "i", getIntEnv(searchIntervalEnv), fmt.Sprintf("Interval to repeat the search. Default: (%v) second", defaultSearchInterval))
 	rootCmd.PersistentFlags().StringVarP(&vaccine, "vaccine", "v", os.Getenv(vaccineEnv), fmt.Sprintf("Vaccine preferences - covishield (or) covaxin. Default: No preference"))
 	rootCmd.PersistentFlags().StringVarP(&fee, "fee", "f", os.Getenv(feeEnv), fmt.Sprintf("Fee preferences - free (or) paid. Default: No preference"))
+	rootCmd.PersistentFlags().StringVarP(&notifier, "notifier", "n", os.Getenv(notifierEnv), "Notifier to use - email (or) telegram. Default: email")
+	rootCmd.PersistentFlags().StringVarP(&tgApiToken, "token", "t", os.Getenv(tgApiTokenEnv), fmt.Sprintf("telegram bot API token"))
+	rootCmd.PersistentFlags().StringVarP(&tgUsername, "username", "u", os.Getenv(tgUsernameEnv), fmt.Sprintf("telegram username"))
```
Comment (PrasadG193) [inline main.go]:
```suggestion
	rootCmd.PersistentFlags().StringVarP(&tgApiToken, "telegram-token", "t", os.Getenv(tgApiTokenEnv), fmt.Sprintf("telegram bot API token"))
	rootCmd.PersistentFlags().StringVarP(&tgUsername, "telegram-username", "u", os.Getenv(tgUsernameEnv), fmt.Sprintf("telegram username"))
```

File: main.go [hunk 2]
```diff
@@ -74,8 +81,17 @@ func checkFlags() error {
...
 	}
-	if len(email) == 0 || len(password) == 0 {
-		return errors.New("Missing email creds")
+	if notifier == "telegram" {
+		if len(tgApiToken) == 0 || len(tgUsername) == 0 {
+			return errors.New("Missing telegram bot api token or username or both")
+		}
+	} else {
+		notifier = "email"
+	}
```
Comment (PrasadG193) [inline main.go]: Let's use switch case here. `default` case can be "email"

File: main.go [hunk 3]
```diff
@@ -13,6 +13,7 @@ import (
 
 var (
 	pinCode, state, district, email, password, date, vaccine, fee string
+	notifier, tgApiToken, tgUsername                              string
```
Comment (PrasadG193) [inline main.go]: Let's declare a interface `Notifier` which implements `Send()` method. We can have a separate package `notifier` for this. Let's refactor the notification workflow

File: pkg/notify/email.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+func (e *Email) SendMessage(body string) error {
+	msg := "From: " + e.Id + "\n" +
+		"To: " + e.Id + "\n" +
+		"Subject: Vaccination slots are available\n\n" +
+		"Vaccination slots are available at the following centers:\n\n" +
+		body
+
+	err := smtp.SendMail("smtp.gmail.com:587",
```
Comment (PrasadG193) [inline pkg/notify/email.go]: Let's declare smtp server address and port as const and use here and in the next line

File: main.go [hunk 5]
```diff
@@ -7,19 +7,25 @@ import (
...
+type NotifierType string
+
 var (
 	pinCode, state, district, email, password, date, vaccine, fee string
-
-	age, interval int
+	tgApiToken, tgUsername                                        string
+	notifierType                                                  NotifierType
+	age, interval                                                 int
+	notifier                                                      notify.Notifier
```
Comment (PrasadG193) [inline main.go]: Let's not declare notifier as a global var. Instead pass it as an arg to checkSlots()

File: main.go [hunk 6]
```diff
@@ -105,10 +127,24 @@ func getIntEnv(envVar string) int {
...
+		return notify.NewEmail(email, password), nil
+	}
+}
+
 func Run(args []string) error {
 	if err := checkFlags(); err != nil {
 		return err
 	}
+	var err error
+	notifier, err = getNotifier()
```
Comment (PrasadG193) [inline main.go]:
```suggestion
	notifier, err := getNotifier()
```
pass `notifier` as an arg ro checkSlots()

File: pkg/notify/telegram.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	Bot    *tgbotapi.BotAPI
+}
+
+func NewTelegram(token, username string) (Notifier, error) {
+	bot, err := tgbotapi.NewBotAPI(token)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Unable to find bot for the given botAPI token %s", token))
+	}
+
+	log.Printf("Authorized on account %s", bot.Self.UserName)
```
Comment (PrasadG193) [inline pkg/notify/telegram.go:34]: Should we return error from here?

File: pkg/notify/telegram.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+		if strings.ToLower(update.Message.Chat.UserName) == strings.ToLower(username) {
+			chatID := update.Message.Chat.ID
+			log.Printf("chatID for the conversation between username: %s and bot: %s is %d", username, bot.Self.UserName, chatID)
+			return &Telegram{
+				ChatID: chatID,
+				Bot:    bot,
+			}, nil
+		}
+	}
+	return nil, errors.New(fmt.Sprintf("Unable to get the chatID \n Send message to the bot %s", bot.Self.UserName))
```
Comment (PrasadG193) [inline pkg/notify/telegram.go]:
```suggestion
	return nil, errors.New(fmt.Sprintf("Unable to get the chatID, send message to the bot %s", bot.Self.UserName))
```

File: pkg/notify/telegram.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+func (t *Telegram) SendMessage(body string) error {
+	if len(body) > 4096 {
+		log.Printf("message body too long, trimming it")
+		body = body[:4095]
+	}
```
Comment (PrasadG193) [inline pkg/notify/telegram.go]: why do we need this?

File: search.go [hunk 10]
```diff
@@ -239,6 +239,7 @@ func getAvailableSessions(response []byte, age int) error {
 		log.Printf("No slots available, rechecking after %v seconds", interval)
 		return nil
 	}
-	log.Print("Found available slots, sending email")
-	return sendMail(email, password, buf.String())
+	log.Print("Found available slots, sending notification")
+	notifier.SendMessage(buf.String())
```
Comment (PrasadG193) [inline search.go]: Should we capture and return error from here?

File: README.md [hunk 11]
```diff
@@ -33,21 +33,25 @@ Usage:
...
+  -s, --state string               Search by state name
+  -v, --vaccine string             Vaccine preferences - covishield (or) covaxin. Default: No preference
+  -n, --notifier-type string       Notifier to use - email (or) telegram. Default: email
+  -t, --telegram-token string      Telegram bot API token
+  -u, --telegram-username string   Telegram username 
 ```
 
 **Note:** Gmail password won't work for 2FA enabled accounts. Follow [this](https://support.google.com/accounts/answer/185833?p=InvalidSecondFactor&visit_id=637554658548216477-2576856839&rd=1) guide to generate app token password and use it with `--password` arg 
 
+**Note:** For creating telegram bot follow [this](https://core.telegram.org/bots#6-botfather). Once the new bot is created successfully, send message to the bot before starting covaccine-notifier.
```
Comment (PrasadG193) [inline README.md]: Could you please create a step-by-step guide with screenshots? We can place it under new dir docs/ and add a link here

File: main.go [hunk 12]
```diff
@@ -7,19 +7,25 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/PrasadG193/covaccine-notifier/pkg/notify"
 	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
```
Comment (PrasadG193) [inline main.go]:
```suggestion
	"github.com/pkg/errors"
	"github.com/spf13/cobra"
	
	"github.com/PrasadG193/covaccine-notifier/pkg/notify"
```
Please import internal packages in separate grp

File: main.go (hunk 5 above)
Comment (PrasadG193) [reply main.go]: I don't think this is resolved

**Comment by PrasadG193** [comment]
I think we should add subcommand for each notifier type instead of adding flags, something like - 
```
covaccine-notifier email --username --password ...
covaccine-notifier telegram --token ...
```

File: pkg/notify/telegram.go (hunk 7 above)
Comment (sohm22) [reply pkg/notify/telegram.go:34]: `if err != nil {
		return nil, errors.New(fmt.Sprintf("Unable to find bot for the given botAPI token %s", token))
}`

we are returning error from here 

`log.Printf("Authorized on account %s", bot.Self.UserName)`

This is just informative log

File: pkg/notify/telegram.go (hunk 9 above)
Comment (sohm22) [reply pkg/notify/telegram.go]: when started notifier using state and district name
I have encountered error for message greater than 4096 characters.

File: search.go (hunk 10 above)
Comment (sohm22) [reply search.go]: yeah. I missed it.  will handle the error

File: pkg/notify/email.go (hunk 4 above)
Comment (sohm22) [reply pkg/notify/email.go]: sure

File: main.go (hunk 6 above)
Comment (sohm22) [reply main.go]: yeah

File: pkg/notify/telegram.go (hunk 7 above)
Comment (PrasadG193) [reply pkg/notify/telegram.go:34]: aah, got it

File: README.md [hunk 13]
```diff
@@ -30,53 +30,70 @@ $ ./covaccine-notifier --help
 CoWIN Vaccine availability notifier India
 
 Usage:
-  covaccine-notifier [FLAGS] [flags]
+  covaccine-notifier [command]
+
+Available Commands:
+  email       Notify slots avaliability using email
```
Comment (PrasadG193) [inline README.md]:
```suggestion
  email       Notify slots availability using email
```

File: README.md [hunk 14]
```diff
@@ -30,53 +30,70 @@ $ ./covaccine-notifier --help
...
+  help        Help about any command
+  telegram    Notify slots availability using telegram
```
Comment (PrasadG193) [inline README.md]:
```suggestion
  telegram    Notify slots availability using Telegram
```

File: main.go [hunk 15]
```diff
@@ -9,17 +9,37 @@ import (
...
 	rootCmd = &cobra.Command{
 		Use:   "covaccine-notifier [FLAGS]",
 		Short: "CoWIN Vaccine availability notifier India",
+	}
+
+	telegramCmd = &cobra.Command{
+		Use:   "telegram [FLAGS]",
+		Short: "Notify slots availability using telegram",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			notifierType = NotifierType(TelegramNotifierType)
```
Comment (PrasadG193) [inline main.go]: I don't think we need to typecast, TelegramNotifierType and EmailNotifierType is already of type NotifierType

File: main.go [hunk 16]
```diff
@@ -46,20 +69,34 @@ const (
 
 	free = "free"
 	paid = "paid"
+
+	EmailNotifierType    NotifierType = "email"
+	TelegramNotifierType NotifierType = "telegram"
```
Comment (PrasadG193) [inline main.go]: Lets move this to notifier package

File: main.go [hunk 17]
```diff
@@ -116,30 +147,42 @@ func getIntEnv(envVar string) int {
...
+		return notify.NewEmail(username, password), nil
+	}
+}
+
 func Run(args []string) error {
 	if err := checkFlags(); err != nil {
 		return err
 	}
-	if err := checkSlots(); err != nil {
+	notifier, err := getNotifier()
```
Comment (PrasadG193) [inline main.go]: We can pass `notify.Notifier` as a arg to Run(). We don't need getNotifier()

File: main.go (hunk 15 above)
Comment (PrasadG193) [reply main.go]: We can just initialize and pass `notifier.Notifier` as an arg to Run from here

File: pkg/notify/email.go [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+	"net/smtp"
+)
+
+const (
+	smtpServerAddress = "smtp.gmail.com"
+	smtpServerPort    = "587"
+)
+
+type Email struct {
+	Id   string
```
Comment (PrasadG193) [inline pkg/notify/email.go]:
```suggestion
	ID   string
```
just a Go way of naming variables :)

File: pkg/notify/notify.go [hunk 19]
```diff
@@ -0,0 +1,5 @@
+package notify
+
+type Notifier interface {
```
Comment (PrasadG193) [inline pkg/notify/notify.go:5]: Please add a comment on exported functions and types. This applies to other changes as well

File: pkg/notify/telegram.go [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+func NewTelegram(username, token string) (Notifier, error) {
+	bot, err := tgbotapi.NewBotAPI(token)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Unable to find bot for the given botAPI token %s", token))
+	}
+
+	log.Printf("Authorized on account %s", bot.Self.UserName)
+
+	u := tgbotapi.NewUpdate(0)
+	u.Timeout = 60
```
Comment (PrasadG193) [inline pkg/notify/telegram.go]: Please declare timeout value as const and use here

File: pkg/notify/telegram.go (hunk 9 above)
Comment (PrasadG193) [reply pkg/notify/telegram.go]: Should we split and send messages into 2 parts? Or send message as a file?

File: search.go [hunk 21]
```diff
@@ -259,6 +261,9 @@ func getAvailableSessions(response []byte, age int, minCapacity int) error {
 		log.Printf("No slots available, min required: %d, rechecking after %v seconds", minCapacity, interval)
 		return nil
 	}
-	log.Print("Found available slots, sending email")
-	return sendMail(email, password, buf.String())
+	log.Print("Found available slots, sending notification")
+	if err := notifier.SendMessage(buf.String()); err != nil {
+		return err
+	}
+	return nil
```
Comment (PrasadG193) [inline search.go]:
```suggestion
	return notifier.SendMessage(buf.String())
```

File: README.md [hunk 22]
```diff
@@ -30,53 +30,70 @@ $ ./covaccine-notifier --help
...
+Use "covaccine-notifier [command] --help" for more information about a command.
+```
+example 
+```
+$ ./covaccine-notifier email --help 
 ```
 
 **Note:** Gmail password won't work for 2FA enabled accounts. Follow [this](https://support.google.com/accounts/answer/185833?p=InvalidSecondFactor&visit_id=637554658548216477-2576856839&rd=1) guide to generate app token password and use it with `--password` arg 
 
+**Note:** For telegram bot integration with covaccine-notifier follow [this](./docs/telegram-integration.md).
```
Comment (PrasadG193) [inline README.md]: Let's add a sub heading - `Integration with Telegram`

File: pkg/notify/telegram.go [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+func (t *Telegram) SendMessage(body string) error {
+	if len(body) > 4096 {
+		log.Printf("Message body too long, Message will be sent as file ")
+		fileBytes := tgbotapi.FileBytes{
+			Name:  fmt.Sprintf("slots-available-%d.txt", time.Now().Unix()),
+			Bytes: []byte(body),
+		}
+		documentConfig := tgbotapi.NewDocumentUpload(t.ChatID, fileBytes)
+		if _, err := t.Bot.Send(documentConfig); err != nil {
+			return errors.Wrap(err, "Unable to send message to telegram")
```
Comment (PrasadG193) [inline pkg/notify/telegram.go:69]: We can return from here. That way we don't need else statement. Put the statements below outside else block

**Review by PrasadG193 — APPROVED** [review]

**Pull request merged**